libs/http2-manager/src/HTTP2/Client/Manager/Internal.hs

```diff
@@ -0,0 +1,337 @@
+import Control.Concurrent.STM
+import Control.Exception
+import Control.Monad
+import Control.Monad.IO.Class
+import Data.ByteString
+import qualified Data.ByteString as BS
+import Data.IORef
+import Data.Map
+import qualified Data.Map as Map
+import Data.Streaming.Network
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Data.Unique
+import Foreign.Marshal.Alloc (mallocBytes)
+import qualified Network.HTTP2.Client as HTTP2
+import qualified Network.Socket as NS
+import qualified OpenSSL.Session as SSL
+import qualified System.TimeManager
+import Prelude
+
+data HTTP2Conn = HTTP2Conn
+  { backgroundThread :: Async (),
+    disconnect :: IO (),
+    -- See comment in 'startPersistentHTTP2Connection' about why this returns
+    -- '()'
+    sendRequestMVar :: MVar (Either Request ())
+  }
+
+type TLSEnabled = Bool
+
+type HostName = ByteString
+
+type Port = Int
+
+type Target = (TLSEnabled, HostName, Port)
+
+-- FUTUREWORK: Support HTTPS, perhaps ALPN negatiation can also be used to
+-- HTTP1. I think HTTP1 vs HTTP2 can not be negotated without TLS, so perhaps
+-- this manager will default to HTTP2.
+data HTTP2Manager = HTTP2Manager
+  { connections :: TVar (Map Target HTTP2Conn),
+    cacheLimit :: Int,
+    sslContext :: SSL.SSLContext
+  }
+
+defaultHTTP2Manager :: IO HTTP2Manager
+defaultHTTP2Manager = do
+  ctx <- SSL.context
+  SSL.contextSetVerificationMode ctx $
+    SSL.VerifyPeer
+      { vpFailIfNoPeerCert = True,
+        -- Only relvant when running as server
+        vpClientOnce = False,
+        vpCallback = Nothing
+      }
+  SSL.contextSetALPNProtos ctx ["h2"]
+  http2ManagerWithSSLCtx ctx
+
+http2ManagerWithSSLCtx :: SSL.SSLContext -> IO HTTP2Manager
+http2ManagerWithSSLCtx sslContext = do
+  connections <- newTVarIO mempty
+  let cacheLimit = 20
+  pure $ HTTP2Manager {..}
+
+-- | Does not check whether connection is actually running. Users should use
+-- 'withHTTP2Request'. This function is good for testing.
+sendRequestWithConnection :: HTTP2Conn -> HTTP2.Request -> (HTTP2.Response -> IO a) -> IO a
+sendRequestWithConnection conn req f = do
+  result <- newEmptyMVar
+  threadKilled <- newEmptyMVar
+  putMVar (sendRequestMVar conn) (Left (Request req (putMVar result <=< f) threadKilled))
+  race (takeMVar result) (takeMVar threadKilled) >>= \case
+    Left x -> pure x
+    Right (SomeException e) -> throw e
+
+-- | Make an HTTP2 request, if it is the first time the 'HTTP2Manager' sees this
+-- (server,port) comibination, it creates the connection and keeps it around for
+-- any subsequent requests. Subsequest requests try to use this connection, in
+-- case the connection is already dead (e.g. the background thread has
+-- finished), a new connection is created.
+--
+-- It is important to consume the response body completely before the
+-- continuation can finish.
+withHTTP2Request :: HTTP2Manager -> TLSEnabled -> ByteString -> Int -> HTTP2.Request -> (HTTP2.Response -> IO a) -> IO a
```

Comment:
use Hostname and Port here.  maybe you should make them newtypes to catch all these.  :-)

Reply:
Newtypes would be annoying, this is going to get stabilized soon, so I think its not necessary. I will find all the places and fix them.